src/redux/api/baseQuery.js

```diff
@@ -0,0 +1,129 @@
+import VersionNumber from 'react-native-version-number';
+import { fetchBaseQuery } from '@reduxjs/toolkit/dist/query/react';
+import { selectBaseURL, selectUser } from '../App/selectors';
+import { Mutex } from 'async-mutex';
+import qs from 'qs';
+import AppUser from '../../AppUser';
+import { logout } from '../App/actions';
+import { setUser } from '../middlewares/HttpMiddleware'
+
+const guestCheckoutEndpoints = [
+  'getOrderValidate',
+  'getOrderTiming',
+  'updateOrder',
+];
+
+const appVersion =
+  VersionNumber.bundleIdentifier +
+  '@' +
+  VersionNumber.appVersion +
+  ' (' +
+  VersionNumber.buildVersion +
+  ')';
+
+// create a new mutex
+const mutex = new Mutex();
+
+const buildBaseQuery = (baseUrl, anonymous = false) => {
+  return fetchBaseQuery({
+    baseUrl,
+    prepareHeaders: (headers, { getState, endpoint }) => {
+      headers.set('X-Application-Version', appVersion);
+
+      if (!anonymous) {
+        const user = selectUser(getState());
+        if (user) {
+          headers.set('Authorization', `Bearer ${user.token}`);
+        } else if (guestCheckoutEndpoints.includes(endpoint)) {
+          //TODO; to be implemented in https://github.com/coopcycle/coopcycle-app/issues/1756
+          // const orderAccessToken = selectOrderAccessToken(getState())
+          //
+          // if (orderAccessToken) {
+          //   headers.set('Authorization', `Bearer ${orderAccessToken}`)
+          // }
+        }
+      }
+
+      return headers;
+    },
+    jsonContentType: 'application/ld+json',
+    timeout: 30000,
+  });
+};
+
+//based on https://redux-toolkit.js.org/rtk-query/usage/customizing-queries#automatic-re-authorization-by-extending-fetchbasequery
+export const baseQueryWithReauth = async (args, api, extraOptions) => {
+  // wait until the mutex is available without locking it
+  await mutex.waitForUnlock();
+
+  const { getState } = api;
+
+  const baseUrl = selectBaseURL(getState()) + '/';
+  const baseQuery = buildBaseQuery(baseUrl);
+
+  let result = await baseQuery(args, api, extraOptions);
+
+  if (result.error && result.error.status === 401) {
+    const user = selectUser(getState());
+
+    if (!user) {
+      return result;
+    }
+
+    const refreshToken = user.refreshToken;
+
+    // checking whether the mutex is locked
+    if (mutex.isLocked()) {
+      // wait until the mutex is available without locking it
```

Comment:
if i understand well we could add the following comment :
`// wait for the mutex release, i.e. wait that another request that needed a fresh token as well gets the new token for us and update the state accordingly`
?